src/MatrixFields/field_name_dict.jl

```diff
@@ -237,6 +399,251 @@ function Base.one(matrix::FieldMatrix)
     return FieldNameDict(inferred_diagonal_keys, entries)
 end
 
+"""
+    get_field_first_index_offset(name::FieldName, ::Type{T}, ::Type{S})
+
+Returns the offset of the the field with name `name` in an object of type `S`
+in multiples of `sizeof(T)`.
+"""
+function get_field_first_index_offset(
+    name::FieldName,
+    ::Type{T},
+    ::Type{S},
+) where {T, S}
+    if name == @name()
+        return 0
+    end
+    child_name = extract_first(name)
+    child_type = fieldtype(S, child_name)
+    remaining_field_chain = drop_first(name)
+    field_index =
+        unrolled_filter(i -> fieldname(S, i) == child_name, 1:fieldcount(S))[1]
+    return DataLayouts.fieldtypeoffset(T, S, field_index) +
+           get_field_first_index_offset(remaining_field_chain, T, child_type)
+end
+if hasfield(Method, :recursion_relation)
+    dont_limit = (args...) -> true
+    for m in methods(get_field_first_index_offset)
+        m.recursion_relation = dont_limit
+    end
+end
+
+function bypass_adjoint(field::Fields.Field)
+    if eltype(field) <: Adjoint
+        return eltype(field.parent)
+    else
+        return eltype(field)
+    end
+end
+
+function bypass_adjoint(d::T) where {T}
+    if T <: Adjoint
+        return typeof(d.parent)
+    else
+        return typeof(d)
+    end
+end
```

Comment:
```suggestion
bypass_adjoint(x::Fields.Field) = bypass_adjoint(Fields.field_values(x))
bypass_adjoint(x::DataLayouts.AbstractData{<:Adjoint}) = eltype(x.parent)
bypass_adjoint(x::DataLayouts.AbstractData) = eltype(x)
bypass_adjoint(x::Adjoint) = typeof(x.parent)
bypass_adjoint(x) = typeof(x)
```
This might be easier to extend?